server/src/main/java/io/crate/planner/optimizer/joinorder/JoinGraph.java

```diff
@@ -191,10 +192,16 @@ public JoinGraph visitJoinPlan(JoinPlan joinPlan, Map<Symbol, LogicalPlan> conte
                     // two keys.
                     if (entry.getKey().size() == 2) {
                         entry.getValue().accept(edgeCollector, context);
+                    } else {
+                        filters.add(entry.getValue());
                     }
                 }
             }
-            return left.joinWith(right).withEdges(edgeCollector.edges);
+            var result = left.joinWith(right).withEdges(edgeCollector.edges);
+            if (filters.isEmpty() == false) {
```

Comment:
minor, just to avoid the if here, but also to avoid doing any work if the list is empty, how about moving this check to the `withFilters()` method?